src/Symfony/Component/DependencyInjection/Compiler/DecoratorServicePass.php

```diff
@@ -66,6 +66,13 @@ public function process(ContainerBuilder $container)
                 $decoratingDefinition = $decoratingDefinitions[$inner];
                 $definition->setTags(array_merge($decoratingDefinition->getTags(), $definition->getTags()));
                 $decoratingDefinition->setTags([]);
+                if ($definition->hasTag('container.service_locator')) {
```

Comment:
Instead of this logic which is very specific, moving `ServiceLocatorTagPass` just before `DecoratorServicePass` in `PassConfig` works too (your integration test remains green).
Can you please have a look?

Reply:
Please double-check #32722 while doing so - another integration test that covers it would be great.

Reply:
Just tried updating my app to Symfony 4.4 and sadly I'm still getting
```
!!  In ServiceLocatorTagPass.php line 52:
!!                                                                                 
!!    Invalid definition for service "App\Shop\MessageHandler\Receiver\ReceiverLo  
!!    cator": an array of references is expected as first argument when the "cont  
!!    ainer.service_locator" tag is set.                                           
```

with my patch everything is working

Reply:
Nevermind my previous comment (deleted it) - once I switched pass order as you suggested everything is fine.

Reply:
As for #32722 I can move `ResolveBindingsPass` to be before `ServiceLocatorTagPass` but I have no clue what it'll cause (and sadly PR you've mentioned added no tests)